packages/mermaid/src/rendering-util/createGradient.ts

```diff
@@ -0,0 +1,747 @@
+
+      // Generate interpolated relative positions within the interval
+      const interpolatedRelativePositions = generateStops(
+        0,
+        relativeHint,
+        1,
+        numTransitionStops,
+        includeStart,
+        false
+      );
+      log.debug(
+        `Generated ${numTransitionStops + (includeStart ? 1 : 0)} interpolated relative positions around the relative transition hint at ${relativeHint} between colors ${currentStop.color === 'HINT' ? prevStop.color : currentStop.color} and ${nextStop.color}: ${interpolatedRelativePositions.join(', ')}`
+      );
+
+      // Append interpolated stops to the gradient
+      interpolatedRelativePositions.forEach((relativePos, hintIndex) => {
+        const absolutePos = absoluteStartPos! + relativePos * totalInterval;
+        const interpolatedColor =
+          hintIndex === 0 && includeStart
+            ? currentStop.color
+            : interpolateColor(startColor, endColor, relativePos, relativeHint);
+        linearGradient
+          .append('stop')
+          .attr('offset', `${absolutePos}%`)
+          .attr('stop-color', interpolatedColor);
+        if (currentStop.color === 'HINT') {
+          log.debug(
+            `Added interpolated transition stop ${hintIndex + 1}/${numTransitionStops} for transition hint between colors ${prevStop.color} and ${nextStop.color} to <linearGradient>: color (interpolated) = '${interpolatedColor}', position = '${absolutePos}%'`
+          );
+        } else {
+          if (hintIndex === 0) {
+            // Treated like a regular stop as this is the starting point of an opacity-gradient-based interpolation
+            log.debug(
+              `Added stop ${index + 1} to <linearGradient>: color = '${interpolatedColor}', position = '${absolutePos}%'`
+            );
+          } else {
+            log.debug(
+              `Added interpolated transition stop ${hintIndex}/${numTransitionStops} for opacity gradient between colors ${currentStop.color} and ${nextStop.color} to <linearGradient>: color (interpolated) = '${interpolatedColor}', position = '${absolutePos}%'`
+            );
+          }
+        }
+      });
+    } else {
+      // Append the standard stop to the gradient with no transition hint or opacity gradient involved in the interpolation
+      linearGradient
+        .append('stop')
+        .attr('offset', `${currentStop.position}%`)
+        .attr('stop-color', currentStop.color);
+      log.debug(
+        `Added stop ${index + 1} to <linearGradient>: color = '${currentStop.color}', position = '${currentStop.position}%'`
+      );
+    }
+  });
+
+  // Setup SVG coordinates
+  const xc = bbox.x + nodeWidth / 2;
+  const yc = bbox.y + nodeHeight / 2;
+  const d1 = (gradientLineLength * (50 - minStop)) / 100;
+  const d2 = (gradientLineLength * (50 - maxStop)) / 100;
+  const { x1, y1, x2, y2 } = { x1: xc, y1: yc + d1, x2: xc, y2: yc + d2 }; // Vertical gradient @ 0deg (bottom to top)
```

Comment:
This is where the magic happens—we use `gradientLineLength` to precisely control how the gradient spans the exact node shape at the given angle, something CSS can only do for rectangular shapes.